### PR TITLE
option to scroll to the top of the next cell

### PR DESCRIPTION
**Pull request by amunger**
for https://github.com/microsoft/vscode/issues/191086

new options for scrolling behavior:

- Scroll to fully reveal the next cell. (current behavior, and using as default for now)
- Scroll to reveal the first line of the next cell. (requested in the linked issue)
- Do not scroll at all. (requested in [this comment](https://github.com/microsoft/vscode-jupyter/issues/13536#issuecomment-1561996110))

After people have been able to experiment, I imagine we will change the default to the second option.

### CODE REVIEW

File: src/vs/workbench/contrib/notebook/browser/view/notebookCellList.ts [hunk 1]
```diff
@@ -917,6 +912,17 @@ export class NotebookCellList extends WorkbenchList<CellViewModel> implements ID
 				}
 				break;
 			case CellRevealPosition.Bottom:
+				if (partial) {
+					const lineHeight = this.viewModel?.layoutInfo?.fontInfo.lineHeight ?? 15;
+					const firstLineLocation = elementTop + lineHeight + 20;
```
Comment (amunger) [inline src/vs/workbench/contrib/notebook/browser/view/notebookCellList.ts:917]: I'm not sure if this extra 20px padding buffer can be calculated more smartly, but it seems to work for all toolbar configurations.

File: src/vs/workbench/contrib/notebook/browser/view/notebookCellList.ts [hunk 2]
```diff
@@ -828,12 +826,12 @@ export class NotebookCellList extends WorkbenchList<CellViewModel> implements ID
 		this._revealInViewWithMinimalScrolling(startIndex);
 	}
 
-	private _revealInViewWithMinimalScrolling(viewIndex: number) {
+	private _revealInViewWithMinimalScrolling(viewIndex: number, partial?: boolean) {
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/view/notebookCellList.ts:829]: do we want to make it explict `firstLine` other than calling it `partial`?

File: src/vs/workbench/contrib/notebook/browser/notebookBrowser.ts [hunk 3]
```diff
@@ -150,7 +150,8 @@ export interface ICommonCellInfo {
 export interface IFocusNotebookCellOptions {
 	readonly skipReveal?: boolean;
 	readonly focusEditorLine?: number;
-	readonly minimalScrolling?: boolean;
+	/** Set a value to only scroll as much as needed, partialReveal will bring the element partially into the view */
+	readonly minimalScrolling?: 'fullReveal' | 'partialReveal' | undefined;
```
[the remark this answers is not in the document]
Comment (rebornix) [reply src/vs/workbench/contrib/notebook/browser/notebookBrowser.ts:154]: I'm wondering if we should align the setting enum and this one, `firstLine` is clear but `partialReveal` is ambigious.

File: src/vs/workbench/contrib/notebook/browser/view/notebookCellList.ts (hunk 1 above)
Comment (amunger) [reply src/vs/workbench/contrib/notebook/browser/view/notebookCellList.ts:917]: I think it's just the space between the top of the cell element, and the top of the first line within the cell.

File: src/vs/workbench/contrib/notebook/test/browser/testNotebookEditor.ts [hunk 4]
```diff
@@ -433,7 +433,6 @@ export function createNotebookCellList(instantiationService: TestInstantiationSe
 		NotebookCellList,
 		'NotebookCellList',
 		DOM.$('container'),
-		viewContext ?? new ViewContext(disposables.add(new NotebookOptions(instantiationService.get(IConfigurationService), instantiationService.get(INotebookExecutionStateService), false)), disposables.add(new NotebookEventDispatcher()), () => ({} as IBaseCellEditorOptions)),
 		delegate,
```
[the remark this answers is not in the document]
Comment (amunger) [reply src/vs/workbench/contrib/notebook/test/browser/testNotebookEditor.ts:436]: Not sure what you mean, I just noticed that the viewContext parameter in notebookCellList was unused, so I removed it.

File: src/vs/workbench/contrib/notebook/browser/notebookBrowser.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (amunger) [reply src/vs/workbench/contrib/notebook/browser/notebookBrowser.ts:154]: unions of strings are far easier to deal with when debugging, but yeah, an enum would be more consistent here

File: src/vs/workbench/contrib/notebook/test/browser/testNotebookEditor.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (amunger) [reply src/vs/workbench/contrib/notebook/test/browser/testNotebookEditor.ts:436]: oh, I can also remove the parameter from that factory function

File: src/vs/workbench/contrib/notebook/browser/view/notebookCellList.ts (hunk 1 above)
Comment (amunger) [reply src/vs/workbench/contrib/notebook/browser/view/notebookCellList.ts:917]: it looks like 6px from the padding-top of .cell-inner-container, 12px from positioning the first line within the cell.
2 more just to give some padding.

Comment (rebornix) [reply src/vs/workbench/contrib/notebook/browser/view/notebookCellList.ts:917]: padding top is `this.viewContext.notebookOptions.getLayoutConfiguration().cellTopMargin`, first line padding is `this.viewContext.notebookOptions.getLayoutConfiguration().editorTopPadding`

**Review by rebornix — CHANGES_REQUESTED** [review]
Let's use `viewContext.notebookOptions.getLayoutConfiguration()` for calculating the padding instead of constants.

**Review by rebornix — DISMISSED** [review]

**Review by rebornix — APPROVED** [review]

**Comment by NickCrews** [comment]
Thank you @amunger, I just got this feature in the newest version and the "do not scroll" option is finally what I have been looking for! 🙏

**Pull request merged**